File: 2022/day-10/day-10.c
/* --- Advent of Code 2023 - Day 10: Cathode-Ray Tube --- */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LENGTH 256

/* store number of cycles, value in register x and final score */
static int no_cycles;
static int x;
static int score;

/* increment the cycle counter and update the score for given cycles */
void tick() {
    if (no_cycles == 20 || no_cycles == 60 || no_cycles == 100 || no_cycles == 140 || no_cycles == 180 || no_cycles == 220) {
        score += no_cycles * x;
    }
    no_cycles++;
}


/* solution part-1 */
int part_1(char input_file[]) {
    FILE *fp = fopen(input_file, "r");
    char buffer[MAX_LENGTH];

    /* inialize variables */
    no_cycles = 1;
    x = 1;
    score = 0;

    /* file opened successfully */
    if (fp != NULL) {
        /* read the opearations and execute them line by line */
        while (fgets(buffer, MAX_LENGTH, fp) != NULL) {
            /* noop takes 1 cycle, addx takes 2 cycles */
            if (!strncmp(buffer, "noop", 4)) {
                tick();
            } else if (!strncmp(buffer, "addx", 4)) {
                tick();
                tick();
                int i = atoi(&(buffer[4]));
                x += i;
            } else {
                fprintf(stderr, "Unknown instruction: %s", buffer);
                exit(1);
            }
        }

        printf("%d\n", score);

    /* file not found */
    } else {
        fprintf(stderr,"Problems opening file '%s'\n", input_file);
        exit(1);
    }

    return 0;
}

/* increment the cycle counter and print a pixel */
void tick_to_print() {
    int printer_pos = no_cycles % 40;

    /* if the x value(+1 / -1) is in range of the printer, a #is printed */
    if (printer_pos == x - 1 || printer_pos == x || printer_pos == x + 1) {
        printf("#");
    }
    else {
        printf(".");
    }

    if (++no_cycles % 40 == 0) {
        printf("\n");
    }
}

/* solution part-2 */
int part_2(char input_file[]) {
    FILE *fp = fopen(input_file, "r");
    char buffer[MAX_LENGTH];

    /* inialize variables */
    no_cycles = 0;
    x = 1;
    score = 0;

    /* file opened successfully */
    if (fp != NULL) {
        while (fgets(buffer, MAX_LENGTH, fp) != NULL) {
            /* noop takes 1 cycle, addx takes 2 cycles */
            if (!strncmp(buffer, "noop", 4)) {
                tick_to_print();
            } else if (!strncmp(buffer, "addx", 4)) {
                tick_to_print();
                tick_to_print();
                int i = atoi(&(buffer[4]));
                x += i;
            } else {
                fprintf(stderr, "Unknown instruction: %s", buffer);
                exit(1);
            }
        }

    /* file not found */
    } else {
        fprintf(stderr,"Problems opening file '%s'\n", input_file);
        exit(1);
    }

    return 0;
}

/* main function */
int main(int argc, char *argv[]) {
    if (argc != 3) {
        fprintf(stderr, "Usage: %s -p[1|2] <input-file>\n", argv[0]);
        return 1;
    }

    if (!strcmp(argv[1], "-p1")) {
        part_1(argv[2]);
    } else if (!strcmp(argv[1], "-p2")) {
        part_2(argv[2]);
    } else {
        fprintf(stderr, "Invalid argument: '%s'\n", argv[1]);
        return 1;
    }

    return 0;
}

File: 2022/day-10/day-10.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

""" --- Advent of Code 2022 - Day 10: Cathode-Ray Tube --- """

import sys

# store number of cycles, value in register x and final score
no_cycles = 0
x = 1
score = 0


# increment the cycle counter and update the score for given cycles
def tick() -> None:
    global no_cycles, x, score
    if no_cycles == 20 or no_cycles == 60 or no_cycles == 100 or no_cycles == 140 or no_cycles == 180 or no_cycles == 220:
        score += no_cycles * x
    no_cycles += 1
    return


# increment the cycle counter and print a pixel
def tick_and_print() -> None:
    global no_cycles, x
    printer_pos = no_cycles % 40

    # if the x value (+1/-1) is in range of the printer, a # is printed
    if printer_pos == x - 1 or printer_pos == x or printer_pos == x + 1:
        print('#', end="")
    else:
        print('.', end="")

    # start a new line every 40 cycles
    no_cycles += 1
    if no_cycles % 40 == 0:
        print("")

    return


# solution part 1
def part_1():
    global no_cycles, x, score

    with open("input.txt") as f:
        # noop takes 1 cycle, addx takes 2 cycles
        for line in f.readlines():
            if line == "noop\n":
                tick()
            elif line[0:4] == "addx":
                tick()
                tick()
                instruction, value = line.split()
                x += int(value)
            else:
                print("Unknown instruction:", line)
                exit(1)

    print(score)


# solution part 2
def part_2():
    global no_cycles, x

    with open("input.txt") as f:
        # noop takes 1 cycle, addx takes 2 cycles
        for line in f.readlines():
            if line == "noop\n":
                tick_and_print()
            elif line[0:4] == "addx":
                tick_and_print()
                tick_and_print()
                instruction, value = line.split()
                x += int(value)
            else:
                print("Unknown instruction:", line)
                exit(1)


# main entry point
if len(sys.argv) > 1 and sys.argv[1] == "-p1":
    part_1()
elif len(sys.argv) > 1 and sys.argv[1] == "-p2":
    part_2()
else:
    print("Usage: python", sys.argv[0], "[-p1|-p2]")
    sys.exit(1)
